src/chapter-03/src/core.rs

```diff
@@ -25,28 +13,30 @@ pub struct ShortenUrlResponse {
     shortened_url: String,
 }
 
-pub struct UrlShortener {}
+#[derive(Default, Debug)]
+pub struct UrlShortener {
+    urls: RwLock<HashMap<String, String>>,
+}
 
 impl UrlShortener {
     pub fn new() -> Self {
-        Self {}
+        Self::default()
     }
 
-    pub fn shorten_url(&self, req: ShortenUrlRequest) -> Result<ShortenUrlResponse, ()> {
+    pub fn shorten_url(&self, req: ShortenUrlRequest) -> ShortenUrlResponse {
```

Comment:
I just realised that here we don't need a `Result` because, right now, this method cannot fail

Reply:
Also, I didn't propose this change, but I was thinking we might simplify the function signature here a little further. Something like:

```rust
pub fn shorten_url(&self, req: &str) -> String {
  // ...
}
```

I like the idea of having DTO-like struct at the edge (controllers), because it can help a lot with validation and with keeping the interface extensible over time, but since here we are operating on an internal helper data structure, we don't need to keep using those.

What do you think?